packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts

```diff
@@ -431,6 +564,117 @@ export function getChangeHandler(
 	return getFieldKind(fieldKinds, kind).changeHandler;
 }
 
+interface CrossFieldTable<TFieldData> {
+	srcTable: NestedMap<RevisionTag | undefined, ChangesetLocalId, unknown>;
+	dstTable: NestedMap<RevisionTag | undefined, ChangesetLocalId, unknown>;
+	srcDependents: NestedMap<RevisionTag | undefined, ChangesetLocalId, TFieldData>;
+	dstDependents: NestedMap<RevisionTag | undefined, ChangesetLocalId, TFieldData>;
+	fieldsToUpdate: Set<TFieldData>;
+}
+
+function newCrossFieldTable<T>(): CrossFieldTable<T> {
+	return {
+		srcTable: new Map(),
+		dstTable: new Map(),
+		srcDependents: new Map(),
+		dstDependents: new Map(),
+		fieldsToUpdate: new Set(),
+	};
+}
+
+interface InvertData {
+	originalRevision: RevisionTag | undefined;
+	fieldChange: FieldChange;
+}
+
+type ComposeData = FieldChange;
+
+interface RebaseData {
+	fieldChange: FieldChange;
+	baseChange: TaggedChange<FieldChangeset>;
+}
+
+interface CrossFieldManagerI<T> extends CrossFieldManager {
+	table: CrossFieldTable<T>;
+	srcQueries: MoveQuerySet;
+	dstQueries: MoveQuerySet;
+	fieldInvalidated: boolean;
+}
+
+function newCrossFieldManager<T>(crossFieldTable: CrossFieldTable<T>): CrossFieldManagerI<T> {
+	const srcQueries = new Map();
+	const dstQueries = new Map();
+	const getMap = (target: CrossFieldTarget) =>
+		target === CrossFieldTarget.Source ? crossFieldTable.srcTable : crossFieldTable.dstTable;
+
+	const getQueries = (target: CrossFieldTarget) =>
+		target === CrossFieldTarget.Source ? srcQueries : dstQueries;
+
+	const manager = {
+		table: crossFieldTable,
+		srcQueries,
+		dstQueries,
+		fieldInvalidated: false,
+		getOrCreate: (
+			target: CrossFieldTarget,
+			revision: RevisionTag | undefined,
+			id: ChangesetLocalId,
+			newValue: unknown,
+		) => {
+			const dependents =
+				target === CrossFieldTarget.Source
+					? crossFieldTable.srcDependents
+					: crossFieldTable.dstDependents;
+			const dependent = tryGetFromNestedMap(dependents, revision, id);
+			if (dependent !== undefined) {
+				crossFieldTable.fieldsToUpdate.add(dependent);
+			}
+
+			if (nestedSetContains(getQueries(target), revision, id)) {
+				manager.fieldInvalidated = true;
+			}
+
+			return getOrAddInNestedMap(getMap(target), revision, id, newValue);
+		},
+		get: (
+			target: CrossFieldTarget,
+			revision: RevisionTag | undefined,
+			id: ChangesetLocalId,
+		) => {
+			addToNestedSet(getQueries(target), revision, id);
+			return tryGetFromNestedMap(getMap(target), revision, id);
+		},
+	};
+
+	return manager;
+}
+
+function finishField<T>(manager: CrossFieldManagerI<T>, fieldData: T) {
```

Comment:
I'm having trouble making sense of the name of this function. Any chance we can give it a more self-contained name?
If there's no obvious better name then let's chat about it once I get a better sense of the rest of the code.